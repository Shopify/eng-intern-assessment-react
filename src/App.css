@import url("https://fonts.googleapis.com/css2?family=Noto+Sans+JP:wght@800&display=swap");

/* Remove default browser margins to color entire page */
body {
  margin: 0px;
  padding: 0px;
}

.App {
  width: 100vw;
  height: 100vh;
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: center;
  background-color: #2f74c0;
  font-family: "Noto Sans JP", sans-serif;
  overflow: hidden;
}

.header {
  text-transform: uppercase;
  font-size: 100px;
  color: white;
  z-index: 1;
  text-align: center;
  padding-top: 2%
}

.laps-container {
  display: flex;
  background-color: #f0f0f0;
  flex-direction: column;
  width: 100%;
  flex-grow: 1;
  overflow: auto;
  animation: slideIn 0.5s ease forwards;
}

.laps-header,
.lap {
  display: flex;
  justify-content: center;
  padding: 10px;
}

.lap-number-header,
.lap-number,
.lap-time-header,
.lap-time {
  flex-basis: 50%;
  text-align: center;
}

.laps-header {
  background-color: #000000;
  color: white;
  font-weight: bold;
  font-size: 30px;
}

.lap {
  margin-top: 10px;
  width: 100%;
  font-size: 25px;
}

.lap-number,
.lap-time {
  margin-right: 20px;
}

@keyframes slideIn {
  from {
    /* Start off-screen from the bottom */
    transform: translateY(100%);
  }
  to {
    /* End at its normal position */
    transform: translateX(0);
  }
}

@media (max-width: 800px) {
  .heading {
    margin: 15px 0;
    font-size: 35px;
  }
}
