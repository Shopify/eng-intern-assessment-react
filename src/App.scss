@use "./styles/partials/global" as *;

.main {
	background-color: #3c3c3b;
	display: flex;
	flex-direction: column;
	align-items: center;
	height: 100vh;
	padding: 10rem auto;
}
