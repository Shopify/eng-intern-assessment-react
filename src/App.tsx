import React from "react";
import Home from "./pages/home";

export default function App() {
  return <Home />;
}
