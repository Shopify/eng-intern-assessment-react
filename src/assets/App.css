@import 'common.css';

body {
  background-color: #eeddd3;
  margin: 0;
  padding: 0;
  font-family: 'Helvetica Neue', sans-serif;
  color: #dcdcdc;
}

.app-container {
  display: flex;
  flex-direction: column;
  align-items: center;
  min-height: 100vh;
  background-color: #eeddd3;
  padding: 20px;
  box-sizing: border-box;
}

/* Responsive Layout */
@media screen and (min-width: 768px) {
  .app-container {
    max-width: 800px;
    margin: 0 auto;
  }
}

.stopwatch-container {
  text-align: center;
}