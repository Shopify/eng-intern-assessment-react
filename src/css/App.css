body {
    margin: 0;
    background-color: #282c34;
}

.stopwatch-container {
    height: 100vh;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
}

.timer-display {
    display: flex;
    color: #ccc;
}

.timer-display p {
    margin: 0;
    font-size: 4rem;
    letter-spacing: 2px; 
}

.timer-display span {
    font-size: 3rem;
    position: relative;
    top: 8px;
    letter-spacing: 2px;
}

.stopwatch-controls-container button {
    margin: 15px;
    border: none;
    background-color: #e5ad3d;
    padding: 3px 12px;
    font-size: 1.3rem;
    font-weight: 400;
    cursor: pointer;
}