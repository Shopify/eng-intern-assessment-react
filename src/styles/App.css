/* Global CSS Variables for Light Theme */
:root {
  --background-color: #fff;
  --text-color: #000;
  --button-bg-color: #34445B;
  --button-text-color: #fff;
  --button-hover-bg-color: #263238;
  --header-bg-color: #f2f2f2;
}

/* Global CSS Variables for Dark Theme */
[data-theme='dark'] {
  --background-color: #121212;
  --text-color: #ffffff;
  --button-bg-color: #1c1c1e;
  --button-hover-bg-color: #313133;
  --header-bg-color: #1c1c1e;
}

/* Base styles for html and body */
html, body {
  margin: 0;
  overflow-y: hidden;
  padding: 0;
  height: 100%;
  background: var(--background-color);
  font-family: "Inter", sans-serif;
}

/* App container styles */
.app {
  position: relative;
  width: 100%;
  min-height: 100vh;
  text-align: center;
  display: flex;
  flex-direction: column;
  align-items: center;
  overflow: hidden;
  color: var(--text-color);
  background-color: var(--background-color);
}

/* Header styles */
.header {
  width: 100%;
  text-align: center;
  background-color: var(--header-bg-color);
  padding-top: 32px;
  padding-bottom: 32px;
}

.header h1 {
  color: var(--text-color);
  font-size: 48px;
  font-family: 'Montserrat', sans-serif;
  margin: 0;
}
