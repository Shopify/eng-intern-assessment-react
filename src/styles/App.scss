@use './breakpoints' as *;
@use './mixins' as *;
@use './fonts' as *;
@use './colours' as *;

* {
    box-sizing: border-box;
    margin: 0;
    font-family: 'Spectral';
  }

  .main {
    
    &_container {
    background-color: $primary-light;
    min-height: 100vh;
    }

    &_heading {
      text-align: center;
      color: $primary-dark;
      font-size: 3.5rem;
      padding: 1rem 0 2rem 0;

      @include tablet {
        padding: 2rem 0;
        font-size: 3rem;
      }
    }
  }
