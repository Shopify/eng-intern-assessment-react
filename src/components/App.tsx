import React, { useState, useEffect, Dispatch, SetStateAction, useRef } from 'react';
import StopWatch from './StopWatch';
import StopWatchButton from './StopWatchButton';
import '../Style.css';


export default function App() {
    const [elapsed, setElapsed] = useState<number>(0);
    const [isRunning, setIsRunning] = useState(false);
    const [lapTimes, setLapTimes] = useState<number[]>([]);
    const [lapStartTime, setLapStartTime] = useState<number | null>(null);

    const lapTimesContainerRef = useRef<HTMLDivElement>(null);

    useEffect(() => {
        let interval: NodeJS.Timeout | undefined;
    
        if (isRunning) {
            const startTime = Date.now() - elapsed;
            interval = setInterval(() => {
                const newElapsed = Date.now() - startTime;
                setElapsed(newElapsed);
            }, 10);
        } else {
            if (interval !== undefined) {
                clearInterval(interval);
            }
        }
    
        return () => {
            if (interval !== undefined) {
                clearInterval(interval);
            }
        };
    }, [isRunning, elapsed, setElapsed]);
    

    const formatTime = (time: number) => {
        const minutes = Math.floor(time / 60000);
        const seconds = Math.floor((time % 60000) / 1000);
        const milliseconds = Math.floor((time % 1000) / 10);

        const pad = (value: number) => (value < 10 ? `0${value}` : value);

        return `${pad(minutes)}:${pad(seconds)}.${pad(milliseconds)}`;
    };

    const handleStartStopClick = () => {
        if (isRunning) {
            setIsRunning(false);
            setLapStartTime(null);
        } else {
            setIsRunning(true);
            if (lapStartTime === null) {
                setLapStartTime(Date.now() - elapsed);
            }
        }
    };

    const handleLapClick = () => {
        if (lapStartTime !== null) {
            const lapTime = Date.now() - lapStartTime;
            setLapTimes([...lapTimes, lapTime]);
            setLapStartTime(Date.now());

            if (lapTimesContainerRef.current) {
                lapTimesContainerRef.current.scrollTop = lapTimesContainerRef.current.scrollHeight;
            }
        }
    };

    const handleResetClick = () => {
        setIsRunning(false);
        setElapsed(0);
        setLapTimes([]);
        setLapStartTime(null);
    };

    return (
        <>
            <StopWatch elapsed={elapsed} formatTime={formatTime} lapTimes={lapTimes} lapTimesContainerRef={lapTimesContainerRef} />
            <StopWatchButton
                isRunning={isRunning}
                handleStartStopClick={handleStartStopClick}
                handleLapClick={handleLapClick}
                handleResetClick={handleResetClick}
            />
        </>
    );
}