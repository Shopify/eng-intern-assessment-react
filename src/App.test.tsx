import React from 'react';
import renderer from 'react-test-renderer';
import App from './App';

describe('App tests', () => {
    it('render should match snapshot file', () => {
        const tree = renderer.create(<App/>).toJSON();
        expect(tree).toMatchSnapshot();
    })
})