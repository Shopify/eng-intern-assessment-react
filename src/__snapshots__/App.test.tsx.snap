// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App tests render should match snapshot file 1`] = `
[
  <div>
    00:00:00.00
  </div>,
  <button
    disabled={false}
    onClick={[Function]}
  >
    Start
  </button>,
  <button
    disabled={true}
    onClick={[Function]}
  >
    Stop
  </button>,
  <button
    disabled={true}
    onClick={[Function]}
  >
    Lap
  </button>,
  <button
    disabled={false}
    onClick={[Function]}
  >
    Reset
  </button>,
]
`;
